### PR TITLE
Test and simplify parseKiloSlashCommands 

### PR DESCRIPTION
**Pull request by markijbema**
Tests mostly added to ensure safe refactoring, but since they don't have external dependencies/mock complexity I'd say they're fine to keep

Change is smaller than it looks, best to look with whitespace disabled commit by commit

### CODE REVIEW

File: src/core/slash-commands/kilo.ts [hunk 1]
```diff
@@ -36,74 +36,66 @@ export async function parseKiloSlashCommands(
 	}
 
 	// this currently allows matching prepended whitespace prior to /slash-command
-	const tagPatterns = [
-		{ tag: "task", regex: /<task>(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*<\/task>/is },
-		{ tag: "feedback", regex: /<feedback>(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*<\/feedback>/is },
-		{ tag: "answer", regex: /<answer>(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*<\/answer>/is },
-		{ tag: "user_message", regex: /<user_message>(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*<\/user_message>/is },
```
Comment (markijbema) [inline src/core/slash-commands/kilo.ts:43]: I noticed in https://github.com/Kilo-Org/kilocode/pull/3988 that we need to keep these 4 in sync, but actually this could just be one regex, eliminating the potential of them getting out of sync

File: src/core/slash-commands/kilo.ts [hunk 2]
```diff
@@ -36,74 +36,47 @@ export async function parseKiloSlashCommands(
...
-	]
+	const tagPattern = /<(task|feedback|answer|user_message)>(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*<\/\1>/is
```
Comment (markijbema) [inline src/core/slash-commands/kilo.ts:39]: Tbc, this is a refactor, but I'd like to change this regex, especially `(\s+.+?)?\s*`; what's happening there. I understand at least one space, but then optionally multiple points, and then possibly some more spaces?

Comment (dltechy) [reply src/core/slash-commands/kilo.ts:39]: I think it's to separate the text after the workflow.
For example: `<task>/hello.md   extra instructions  </task>`
- `\s+.+?` handles the initial space and the extra words (`   extra instructions`)
  - `.+?` is any number of character, not points since it's not inside `[]`.
- The final `\s*` handles the trailing spaces if they exist.

Comment (markijbema) [reply src/core/slash-commands/kilo.ts:39]: oh you're right it's way to non-greedy-match everything except the trailing whitespace. I really dislike non greedy matching in regexes :P

File: src/core/slash-commands/kilo.ts (hunk 1 above)
Comment (EamonNerbonne) [reply src/core/slash-commands/kilo.ts:43]: If you prefer, you could instead extract the sub-regex (prob as a substring) `(\s*\/([a-zA-Z0-9_.-]+))(\s+.+?)?\s*` and reuse that; the sole advantage being avoiding the backref, I guess?

File: src/core/slash-commands/kilo.ts [hunk 3]
```diff
@@ -36,74 +36,47 @@ export async function parseKiloSlashCommands(
...
 
-	// if we find a valid match, we will return inside that block
-	for (const { regex } of tagPatterns) {
-		const regexObj = new RegExp(regex.source, regex.flags)
-		const match = regexObj.exec(text)
+	const match = tagPattern.exec(text)
```
Comment (EamonNerbonne) [inline src/core/slash-commands/kilo.ts:41]: I doubt it matters, since we probably don't have multiple of these, but the order will now be in string-position order rather than in order of the tagPatterns array.

File: src/core/slash-commands/kilo.ts [hunk 4]
```diff
@@ -36,74 +36,47 @@ export async function parseKiloSlashCommands(
...
-			// match[2] is just the command name (e.g. "newtask")
+	if (match) {
+		const [fullMatch, _tagName, commandWithWhitespace, commandName] = match
+		const fullMatchStartIndex = match.index
+		const command = commandReplacements[commandName]
 
-			const commandName = match[2] // casing matters
-			const command = commandReplacements[commandName]
+		// find position of slash command within the full match
+		const relativeStartIndex = fullMatch.indexOf(commandWithWhitespace)
```
Comment (EamonNerbonne) [inline src/core/slash-commands/kilo.ts:49]: isn't this match.indices[2][0]?

Comment (EamonNerbonne) [reply src/core/slash-commands/kilo.ts:49]: oh, only with the d flag in the regex.  🙄

Comment (EamonNerbonne) [reply src/core/slash-commands/kilo.ts:49]: ...and then it'd be `match.indices[2][0] - match.index` since its relative to the match, not the string.

**Comment by EamonNerbonne** [comment]
A nice sanity check would be to merge the test first, and make sure it contains enough wacko corner cases (can we nest these tags or have multiple?)

Or at least have a git commit with just the test be green, then change the implementation.

I mean, it looks sane, but hard to say we're not missing anything this way...

File: src/core/slash-commands/kilo.ts (hunk 1 above)
Comment (markijbema) [reply src/core/slash-commands/kilo.ts:43]: sure, still seems silly to have multiple regexes for this though?

File: src/core/slash-commands/kilo.ts (hunk 3 above)
Comment (markijbema) [reply src/core/slash-commands/kilo.ts:41]: I'm pretty sure this always matches the whole message

**Comment by markijbema** [comment]
Triggering: https://github.com/Kilo-Org/kilocode/pull/4048

I did run them myself of course

File: src/core/slash-commands/kilo.ts (hunk 4 above)
Comment (markijbema) [reply src/core/slash-commands/kilo.ts:49]: I havent really looked at/changed the inner algorithm, other than giving the matches a name. I mean, if we're already parsing with regex we can probably also match textWithoutSlashCommand with te regex, but i was trying to be careful :P

**Review by EamonNerbonne — APPROVED** [review]
